packages/astro/src/runtime/server/render/astro/instance.ts

```diff
@@ -0,0 +1,82 @@
+import type { SSRResult } from '../../../../@types/astro';
+import type { AstroComponentFactory, AstroFactoryReturnValue } from './factory.js';
+
+import { HydrationDirectiveProps } from '../../hydration.js';
+import { renderChild } from '../any.js';
+import { isHeadAndContent } from './head-and-content.js';
+import { isAPropagatingComponent } from './factory.js';
+import { isPromise } from '../../util.js';
+
+type ComponentProps = Record<string | number, any>;
+
+const astroComponentInstanceSym = Symbol.for('astro.componentInstance');
+
+export class AstroComponentInstance {
+	[astroComponentInstanceSym] = true;
+
+	private readonly result: SSRResult;
+	private readonly props: ComponentProps;
+	private readonly slots: any;
+	private readonly factory: AstroComponentFactory;
+	private returnValue: ReturnType<AstroComponentFactory> | undefined;
+	constructor(result: SSRResult, props: ComponentProps, slots: any, factory: AstroComponentFactory) {
+		this.result = result;
+		this.props = props;
+		this.slots = slots;
+		this.factory = factory;
+	}
+
+	async init() {
+		this.returnValue = this.factory(this.result, this.props, this.slots);
+		return this.returnValue;
+	}
+
+	async *render() {
+		if(this.returnValue === undefined) {
+			await this.init();
+		}
+
+		let value: AstroFactoryReturnValue | undefined = this.returnValue;
+		if(isPromise(value)) {
+			value = await value;
+		}
+		if(isHeadAndContent(value)) {
+			yield * value.content;
+		} else {
+			yield * renderChild(value);
+		}
+	}
+}
+
+// In dev mode, check props and make sure they are valid for an Astro component
+function validateComponentProps(props: any, displayName: string) {
+	if (import.meta.env?.DEV && props != null) {
```

Comment:
Merge note: `import.meta.env?.DEV` was removed in https://github.com/withastro/astro/pull/5501

Reply:
Oh ok, how do we know we are running in dev now?

Reply:
The idea was that the warning should be called on build too. Does this need to be dev only?

Reply:
Build is ok, but in SSR this is going to be called on every render which is quite expensive.